### ISSUE TITLE
zfs hangs on suspend/resume

### ISSUE THREAD

**Issue by gregfr** [zfsonlinux/zfs#4701]
Greetings

I've created a zfs pool on an external drive. Then I put the system to suspend.

A second after going to suspend, it resumed. Now every command related to zfs or the mounted volumes just hangs (including `zpool list`, `zfs list`, `zpool status`...). I've even tried `zpool clear -nFX AUX`. The only command not hanging is "`zfs umount`" which gives:

`cannot open 'AUX': pool I/O is currently suspended`

After some googling, I found out that the disk was "relinked" (not sure what is the right word) from /dev/sdj to /dev/sde; however I've used partlabel to create my pool (not dev name).

I can make the disk spin up with `blkid`, after a while it spins down.

How can I unblock all the process and go on with the disk?

Thanks in advance

I'm using Ubuntu kernel 3.5.0-51-generic - I don't know how to make zfs tell me its version

```
$ modinfo zfs
filename:       /lib/modules/3.5.0-51-generic/updates/dkms/zfs.ko
license:        CDDL
author:         Sun Microsystems/Oracle, Lawrence Livermore National Laboratory
description:    ZFS
srcversion:     6B0A2BDB1B0D0798F123A83
depends:        znvpair,spl,zcommon,zunicode,zavl
vermagic:       3.5.0-51-generic SMP mod_unload modversions 
```

**Comment by gregfr** [comment]
I think zfs/spl versions are 0.6.2:

```
dpkg --list | grep zfs
ii  dkms                                                                 2.2.0.3-1.1ubuntu1.1+zfs6~quantal1                all          Dynamic Kernel Module Support Framework
ii  libzfs1                                                              0.6.2-1~quantal                                   amd64        Native ZFS filesystem library for Linux
ii  mountall                                                             2.42ubuntu0.2-zfs1                                amd64        filesystem mounting tool
ii  ubuntu-zfs                                                           7~quantal                                         amd64        Native ZFS filesystem metapackage for Ubuntu.
ii  zfs-dkms                                                             0.6.2-1~quantal                                   amd64        Native ZFS filesystem kernel modules for Linux
ii  zfsutils                                                             0.6.2-1~quantal                                   amd64        Native ZFS management utilities for Linux

```

**Comment by behlendorf** [comment]
@gregfr I'd strongly recommend updating to the latest release v0.6.5.7.  It's available through the Ubuntu PPA or as of 16.04 from the main Ubuntu repositories.  It contains many important bug fixes and one which might address this issue.

**Comment by gregfr** [comment]
I cannot update ubuntu, can I still update zfs? should I compile it manually?

**Comment by behlendorf** [comment]
You should still be able to update ZFS from the PPA for 14.04.

**Comment by gregfr** [comment]
Is it here? https://launchpad.net/~zfs-native/+archive/ubuntu/stable
I'm running ubuntu 12, not 14...

**Comment by behlendorf** [comment]
Yes, that's the spot it looks like 0.6.5.4-1 is available for precise which is a lot newer than what your running.

**Comment by gregfr** [comment]
Unfortunalty I'm running quantal not precise. Maybe I can un-install the packages and compile it from sources?

**Comment by gregfr** [comment]
If I do : `./configure --with-spl=../spl-0.6.5.7 `, configure runs ok but zfs doesn't compile. If I do : `./configure --with-spl=../spl-0.6.5.7 --with-spl-obj=../spl-0.6.5.7/module/spl/' configure gives me an error... what is the right path for "--with-spl-obj"?

**Comment by gregfr** [comment]
Am I missing something obvious?

**Comment by behlendorf** [comment]
@gregfr the `--with-spl*=` options don't currently handle relative paths, try using an absolute one.  It's not obvious.

**Comment by gregfr** [comment]
I forgot to thank you for this reply :)
I was able to compile, but the test scripts seems to be missing?

**Comment by behlendorf** [comment]
The ZFS Test Suite was added after the 0.6.5.x releases.

**Comment by qyra** [comment]
Was this issue ever fixed? I have zfs version 0.7.5-1ubuntu16.4, running zfs on Ubuntu 18.04.2 LTS, and the exact same issue is happening to me. I've got ZFS running on an external USB drive, and if I ever suspend my system while it is mounted then after I resume zfs will hang forever with IO errors (It also won't let me export and re-import, the only fix I found was to restart my system entirely)

I also tried setting up a bash script to export the pool before sleep, but this won't work because if anything is using the pool before suspend then the export will fail.
